File: src/java/com/portalpostal/dao/handler/LancamentoProgramadoHandler.java
package com.portalpostal.dao.handler;

import com.portalpostal.model.Conta;
import com.portalpostal.model.LancamentoParcelado;
import com.portalpostal.model.LancamentoProgramado;
import com.portalpostal.model.PlanoConta;
import com.portalpostal.model.TipoDocumento;
import com.portalpostal.model.TipoFormaPagamento;
import com.portalpostal.model.dd.TipoFrequencia;
import com.portalpostal.model.dd.TipoLancamento;
import com.portalpostal.model.dd.TipoSituacaoLancamentoProgramado;
import java.sql.ResultSet;
import java.sql.SQLException;
import org.sql2o.ResultSetHandler;

public class LancamentoProgramadoHandler extends GenericHandler implements ResultSetHandler<LancamentoProgramado> {
    
    public LancamentoProgramadoHandler() {
        super("lancamento_programado");
    }
    
    public LancamentoProgramadoHandler(String table) {
        super(table);
    }

    public LancamentoProgramado handle(ResultSet result) throws SQLException {
        LancamentoProgramado lancamentoProgramado = new LancamentoProgramado();
        lancamentoProgramado.setIdLancamentoProgramado(getInt(result, "idLancamentoProgramado"));
        lancamentoProgramado.setConta(getConta(result));
        lancamentoProgramado.setPlanoConta(getPlanoConta(result));
        lancamentoProgramado.setLancamentoParcelado(getLancamentoParcelado(result));
        lancamentoProgramado.setTipo(TipoLancamento.values()[getInt(result, "tipo")]);
        lancamentoProgramado.setFavorecido(getString(result, "favorecido"));
        lancamentoProgramado.setNumero(getString(result, "numero"));
        lancamentoProgramado.setDocumento(getTipoDocumento(result));
        lancamentoProgramado.setFormaPagamento(getTipoFormaPagamento(result));
        lancamentoProgramado.setFrequencia(TipoFrequencia.values()[getInt(result, "frequencia")]);
        lancamentoProgramado.setNumeroParcela(getInt(result, "numeroParcela"));
        lancamentoProgramado.setDataEmissao(getDate(result, "dataEmissao"));
        lancamentoProgramado.setDataVencimento(getDate(result, "dataVencimento"));
        lancamentoProgramado.setValor(getDouble(result, "valor"));
        lancamentoProgramado.setSituacao(TipoSituacaoLancamentoProgramado.values()[getInt(result, "situacao")]);
        lancamentoProgramado.setHistorico(getString(result, "historico"));
        return lancamentoProgramado;
    }
    
    private Conta getConta(ResultSet result) throws SQLException {
        if(!existColumn(result, "conta.idConta")) return null;
        return new ContaHandler().handle(result); 
    }
    
    private PlanoConta getPlanoConta(ResultSet result) throws SQLException {
        if(!existColumn(result, "plano_conta.idPlanoConta")) return null;
        return new PlanoContaHandler().handle(result); 
    }
    
    private LancamentoParcelado getLancamentoParcelado(ResultSet result) throws SQLException {
        if(!existColumn(result, "lancamento_parcelado.idLancamentoParcelado")) return null;
        if(!existFKValue(result, "lancamento_parcelado.idLancamentoParcelado")) return null;
        return new LancamentoParceladoHandler().handle(result); 
    }
    
    private TipoDocumento getTipoDocumento(ResultSet result) throws SQLException {
        if(!existColumn(result, "tipo_documento.idTipoDocumento")) return null;
        return new TipoDocumentoHandler().handle(result); 
    }
    
    private TipoFormaPagamento getTipoFormaPagamento(ResultSet result) throws SQLException {
        if(!existColumn(result, "tipo_forma_pagamento.idTipoFormaPagamento")) return null;
        return new TipoFormaPagamentoHandler().handle(result); 
    }
    
}

File: web/app/js/handler/lancamentoProgramadoHandler.js
'use strict';

app.factory('LancamentoProgramadoHandler', function() {
    
    var handleList = function(lancamentoProgramadoList) { 
        return _.map(lancamentoProgramadoList, function(lancamentoProgramado) {
            return handle(lancamentoProgramado);
        });
    };
    
    var handle = function(lancamentoProgramado) {  
        var lancamentoProgramadoHandle = {};
        lancamentoProgramadoHandle.idLancamentoProgramado = getId(lancamentoProgramado);            
        lancamentoProgramadoHandle.conta = getConta(lancamentoProgramado);  
        lancamentoProgramadoHandle.planoConta = getPlanoConta(lancamentoProgramado);  
        lancamentoProgramadoHandle.centroCusto = getCentroCusto(lancamentoProgramado);
        lancamentoProgramadoHandle.tipo = getTipo(lancamentoProgramado); 
        lancamentoProgramadoHandle.favorecido = getFavorecido(lancamentoProgramado); 
        lancamentoProgramadoHandle.numero = getNumero(lancamentoProgramado); 
        lancamentoProgramadoHandle.documento = getDocumento(lancamentoProgramado); 
        lancamentoProgramadoHandle.formaPagamento = getFormaPagamento(lancamentoProgramado); 
        lancamentoProgramadoHandle.frequencia = getFrequencia(lancamentoProgramado);
        lancamentoProgramadoHandle.quantidadeParcela = getQuantidadeParcela(lancamentoProgramado);
        lancamentoProgramadoHandle.numeroParcela = getNumeroParcela(lancamentoProgramado);
        lancamentoProgramadoHandle.dataCompetencia = getDataCompetencia(lancamentoProgramado);
        lancamentoProgramadoHandle.dataEmissao = getDataEmissao(lancamentoProgramado);
        lancamentoProgramadoHandle.dataVencimento = getDataVencimento(lancamentoProgramado);
        lancamentoProgramadoHandle.valor = getValor(lancamentoProgramado);
        lancamentoProgramadoHandle.situacao = getSituacao(lancamentoProgramado);
        lancamentoProgramadoHandle.historico = getHistorico(lancamentoProgramado);
        lancamentoProgramadoHandle.usuario = getUsuario(lancamentoProgramado);
        return lancamentoProgramadoHandle;
    };
    
    var getId = function(lancamentoProgramado) {
        return lancamentoProgramado.idLancamentoProgramado || null;
    };
    
    var getConta = function(lancamentoProgramado) {
        if(!lancamentoProgramado.conta) return lancamentoProgramado.conta;
        return { idConta: lancamentoProgramado.conta.idConta };
    };
    
    var getPlanoConta = function(lancamentoProgramado) {
        if(!lancamentoProgramado.planoConta) return lancamentoProgramado.planoConta;    
        return { idPlanoConta: lancamentoProgramado.planoConta.idPlanoConta }; 
    };
    
    var getCentroCusto = function(lancamentoProgramado) {
        if(!lancamentoProgramado.centroCusto) return lancamentoProgramado.centroCusto;    
        return { idCentroCusto: lancamentoProgramado.centroCusto.idCentroCusto }; 
    };
    
    var getTipo = function(lancamentoProgramado) {
        if(!lancamentoProgramado.tipo) return lancamentoProgramado.tipo.id;
        return lancamentoProgramado.tipo.id; 
    };
    
    var getFavorecido = function(lancamentoProgramado) {
        return lancamentoProgramado.favorecido || null;
    };
    
    var getNumero = function(lancamentoProgramado) {
        return lancamentoProgramado.numero;
    };
    
    var getDocumento = function(lancamentoProgramado) {
        if(!lancamentoProgramado.documento) return lancamentoProgramado.documento.idTipoDocumento;
        return { idTipoDocumento: lancamentoProgramado.documento.idTipoDocumento };
    };
    
    var getFormaPagamento = function(lancamentoProgramado) {
        if(!lancamentoProgramado.formaPagamento) return lancamentoProgramado.formaPagamento.idTipoFormaPagamento;
        return { idTipoFormaPagamento: lancamentoProgramado.formaPagamento.idTipoFormaPagamento };
    };
    
    var getFrequencia = function(lancamentoProgramado) {
        if(!lancamentoProgramado.frequencia) return lancamentoProgramado.frequencia.id;
        return lancamentoProgramado.frequencia.id;
    };
    
    var getQuantidadeParcela = function(lancamentoProgramado) {
        return lancamentoProgramado.quantidadeParcela;
    };
    
    var getNumeroParcela = function(lancamentoProgramado) {
        return lancamentoProgramado.numeroParcela;
    };
    
    var getDataCompetencia = function(lancamentoProgramado) {
        return lancamentoProgramado.dataCompetencia || moment(); 
    };
    
    var getDataEmissao = function(lancamentoProgramado) {
        return lancamentoProgramado.dataEmissao || moment(); 
    };
    
    var getDataVencimento = function(lancamentoProgramado) {
        return lancamentoProgramado.dataVencimento || lancamentoProgramado.dataLancamento || moment(); 
    };
    
    var getValor = function(lancamentoProgramado) {
        return lancamentoProgramado.valor || 0;
    };
    
    var getSituacao = function(lancamentoProgramado) {
        if(!lancamentoProgramado.situacao) return lancamentoProgramado.situacao.id;
        return lancamentoProgramado.situacao.id; 
    };
    
    var getHistorico = function(lancamentoProgramado) {
        return lancamentoProgramado.historico || null;
    };
    
    var getUsuario = function(lancamentoProgramado) {
        return lancamentoProgramado.usuario || null;
    };

    return {
        handle: handle,
        handleList: handleList
    }

});